tests/kernel/semaphore/semaphore/src/main.c

```diff
@@ -888,24 +1145,48 @@ void test_sem_multi_take_timeout_diff_sem(void)
 
 }
 
+/**
+ * @brief Test thread mutual exclusion by semaphore
+ * @details Test is using to see how mutual exclusion is made by semaphore
+ * @ingroup kernel_semaphore_tests
+ */
+void test_sem_queue_mutual_exclusion(void)
+{
+	k_sem_reset(&simple_sem);
+	k_sem_give(&simple_sem);
+
+	k_thread_create(&sem_tid_1, stack_1, STACK_SIZE,
+			sem_queue_mutual_exclusion, NULL, NULL,
+			NULL, K_PRIO_PREEMPT(-1), 0,
+			K_NO_WAIT);
+
+	sem_queue_mutual_exclusion(NULL, NULL, NULL);
```

Comment:
I don't know how to split that patch correctly, because I made it in a single commit. Seems will be necessary to remove all non-logic things changes, commit only with changes for the tests, then add changes of the non-logic things, make second commit...

That test test_sem_queue_mutual_exclusion I wanted to make sure that kernel can provide a counting semaphore abstraction for queuing, and also for mutual exclusion. Seems that I did that wrong, will rework that test.

Reply:
I don't know how to split that patch correctly, because I made it in a single commit. Seems will be necessary to remove all non-logic things changes, commit only with changes for the tests, then add changes of the non-logic things, make second commit...

Reply:
I don't want to enforce busywork for busywork's sake, but to be honest: patch splitting/squashing and rebasing is one of those git tricks that you're going to need to learn at some point anyway. :)

Basically, make this patch your HEAD, "`git reset HEAD^`", then "`git add -p`" and answer the questions as they come up as to whether or not you want to add the current chunk to the index.  Say "y" to all the ones you want to save in a single commit, then "`git commit`" to create a new commit with just the index.  Then repeat.

Reply:
Ok, thanks for that tip! I didn't know about that.

Reply:
@andyross I reworked mutual exclusion test, please take a look, then I will start to split changes.